### PR TITLE
Fix the `--serial` and `--fail-fast` flags.

### LINKED ISSUE #499: Integration test for `serial` option
We should have an integration test for the public API to make sure we don't break serial test running. It has happened before.

https://github.com/sindresorhus/ava/blob/master/test/api.js

It should ensure tests are not run concurrently.

### PR DESCRIPTION
**Pull request by jamestalmage**
Also fixes the test that was supposed to cover `--serial`.

Fixes #499

### CODE REVIEW

**Comment by jamestalmage** [comment]
I am not really sure why the last commit is more performant. It would obviously be more performant to extend the prototype if we were creating many Runner instances, but we only create one per process.

Two comparisons before the last commit:

screenshot 2016-02-02 16 35 56
screenshot 2016-02-02 16 35 23

after the last commit:

screenshot 2016-02-02 16 49 16

File: index.js [hunk 1]
```diff
@@ -10,8 +10,11 @@ var log = require('./lib/logger');
 // note that test files have require('ava')
 require('./lib/test-worker').avaRequired = true;
 
-var opts = JSON.parse(process.argv[2]);
-var runner = new Runner(opts);
+var opts = globals.options;
+var runner = new Runner({
+	serial: Boolean(opts.serial),
+	bail: Boolean(opts.failFast)
```
Comment (sindresorhus) [inline index.js]: Why do you need to coerce these to booleans? If it's needed it should happen at the source where they're gotten/defined.

File: test/runner.js [hunk 2]
```diff
@@ -209,3 +209,111 @@ test('only test', function (t) {
 		t.end();
 	});
 });
+
+test('options.serial forces all tests to be serial', function (t) {
+	t.plan(1);
```
Comment (sindresorhus) [inline test/runner.js]: add an empty line below `t.plan(1);`

File: lib/test-collection.js [hunk 3]
```diff
@@ -121,7 +121,8 @@ function buildTestWithHooks(entry, beforeEach, afterEach, report) {
 	return arr.concat(buildHooks(afterEach, entry.title, contextRef, report));
 }
 
-TestCollection.prototype.buildPhases = function (report) {
+TestCollection.prototype.buildPhases = function (report, bail) {
+	bail = Boolean(bail);
```
Comment (sindresorhus) [inline lib/test-collection.js]: Same here. Can you coerce it to a boolean at the source?

**Comment by sindresorhus** [comment]
:+1: Generally looks good :)

File: test/runner.js (hunk 2 above)
Comment (jamestalmage) [inline test/runner.js]: done

**Comment by jamestalmage** [comment]
@sindresorhus - made all the recommended changes except the `global.Promise = ` one.

**Comment by sindresorhus** [comment]
You missed https://github.com/sindresorhus/ava/pull/503/files#r51689060 ;)

**Comment by jamestalmage** [comment]
OK. Done now.

**Comment by sindresorhus** [comment]
Thanks for fixing so quickly :)

**Pull request closed without merging**